### PR TITLE
debugger: proper readline support, step.3

### LINKED ISSUE #7738: debugger followup: make debugger REPL compatible with Tarantool console
After the debugger.1 patch (#7593) we have situation when we could not load debugger from inside of Tarantol console REPL due to incompatibility with their readline implementation.
```
tarantool> dbg = require 'luadebug'
---
...

tarantool> dbg() << here we will hung

```

- we should reuse Tarantool console readline functionality for that (i.e. extracting the basic module into `internal.console` for example, and then reusing it both in `console.lua` and `luadebug.lua`);
- make sure debugger luatest is not fragile due to unpredictable popen behavior;

### PR DESCRIPTION
**Pull request by tsafin**
Implemented readline history and autocomplete via reusing of Tarantool console readline facilities. They were hidden once Lua 'console' module loaded, now they kept visible as 'console_internal'.

Followup #7593 
Closes #7738

### CODE REVIEW

File: changelogs/unreleased/gh-7593-console-debugger.md [hunk 1]
```diff
@@ -2,20 +2,5 @@
...
->...
->
->tarantool> dbg()
->---
->```
-> Users should call debugger activation only in their instrumented code, not
-> from the interactive console.
-
 * Introduced a new Lua API `tarantool.debug.getsources()` which allows
   seeing sources of builtin modules in any external debugger.
```
Comment (andreyaksenov) [inline changelogs/unreleased/gh-7593-console-debugger.md:6]: Minor suggestions regarding changelog (for example, builtin -> built-in):
```
* Introduced a new console debugger `luadebug.lua` for debugging external and
  built-in Lua modules.
* Introduced a new Lua API `tarantool.debug.getsources()`, which allows
  seeing sources of built-in modules in any external debugger.
```

**Review by andreyaksenov — APPROVED** [review]

**Review by p7nov — APPROVED** [review]
Please apply the changelog suggestion

File: changelogs/unreleased/gh-7593-console-debugger.md (hunk 1 above)
Comment (Lord-KA) [reply changelogs/unreleased/gh-7593-console-debugger.md:6]: Applied.

File: third_party/lua/luadebug.lua [hunk 2]
```diff
@@ -27,6 +27,7 @@
 local dbg
 
 local DEBUGGER = 'luadebug'
+local HISRTORYFILE = '.tdbg_history'
```
Comment (Totktonada) [inline third_party/lua/luadebug.lua:30]: Nit: Should it actually be named as '**hisrtory** file'? :)

File: test/app-luatest/console_debugger_session_test.lua [hunk 3]
```diff
@@ -298,25 +298,38 @@ local function run_debug_session(cmdline, sequence, tmpfile, header)
...
-                debuglog('stdout output:', result)
-            until result ~= '' or result ~= clean_cmd
-            if expected ~= '' then
-                t.assert_str_contains(result, expected, false)
+
+            -- The output is returned in several chunks, first the prompt,
+            -- then the command, and finally the result. In most cases it
+            -- is can be read in one go, but to prevent flacking the reading
+            -- may be retried.
+            if (expected ~= '') then
```
Comment (Totktonada) [inline test/app-luatest/console_debugger_session_test.lua:324]: Nit: Redundant parentheses.

File: test/app-luatest/console_debugger_session_test.lua [hunk 4]
```diff
@@ -298,25 +298,38 @@ local function run_debug_session(cmdline, sequence, tmpfile, header)
...
+                t.helpers.retrying({timeout = 3, delay = 0.05},
+                function(expected)
+                    local result = trim(unescape(fh:read({ timeout = 0.5 })))
+                    t.assert_str_contains(result, expected, false)
+                    debuglog('stdout output:', result)
+                end, expected)
+            else
+                fh:read({ timeout = 0.5 })
```
Comment (Totktonada) [inline test/app-luatest/console_debugger_session_test.lua:332]: What is purpose of this `:read()`?

File: test/app-luatest/console_debugger_session_test.lua [hunk 5]
```diff
@@ -298,25 +298,38 @@ local function run_debug_session(cmdline, sequence, tmpfile, header)
...
+            -- then the command, and finally the result. In most cases it
+            -- is can be read in one go, but to prevent flacking the reading
+            -- may be retried.
+            if (expected ~= '') then
+                t.helpers.retrying({timeout = 3, delay = 0.05},
+                function(expected)
+                    local result = trim(unescape(fh:read({ timeout = 0.5 })))
+                    t.assert_str_contains(result, expected, false)
+                    debuglog('stdout output:', result)
+                end, expected)
```
Comment (Totktonada) [inline test/app-luatest/console_debugger_session_test.lua:330]: It doesn't solve the problem from the comment above:

> The output is returned in several chunks

For example, you're looking for `foo`. The first chunk is `fo`, the second is `o`. This code tries to find `foo` in `fo` and `o` instead of looking into the concatenated result.

File: test/app-luatest/console_debugger_session_test.lua [hunk 6]
```diff
@@ -298,25 +298,38 @@ local function run_debug_session(cmdline, sequence, tmpfile, header)
...
-            local result
-            local clean_cmd = trim(cmd)
-            -- there should be empty stderr - check it before stdout
-            local errout = fh:read({ timeout = 0.05, stderr = true})
-            debuglog('stderr output: ', trim(errout))
-            if first and errout then
-                -- we do not expect anything on stderr
-                -- with exception of initial debugger header
-                t.assert_str_contains(trim(errout), header, false)
+            if first then
```
Comment (Totktonada) [inline test/app-luatest/console_debugger_session_test.lua:301]: I see no direct relation of the test changes to the code changes. Please, either extract it into its own commit with a proper description of the solved problem or make this relation clear in the commit message.

File: third_party/lua/luadebug.lua [hunk 7]
```diff
@@ -97,11 +98,19 @@ local stack_top = 0
...
 local stack_inspect_offset = 0
 
--- Default dbg.read function
-local function dbg_read(prompt)
-    dbg.write(prompt)
-    io.flush()
-    return io.read()
+-- Tarantool console compatible readline support.
+local function dbg_readline(prompt)
+    local console = require 'console.lib'
```
Comment (Totktonada) [inline third_party/lua/luadebug.lua:103]: From the commit message:

> now they kept visible as 'console_internal'.

Does you mean `console.lib`?

Comment (Totktonada) [reply third_party/lua/luadebug.lua:103]: I would add that it is changed in c432e9e918f1caa69a3e6050830c9f2734217ba7.

File: test/app-luatest/console_debugger_session_test.lua (hunk 4 above)
Comment (Lord-KA) [reply test/app-luatest/console_debugger_session_test.lua:332]: To clear out the buffer for the followup reads. While we don't need to check the output in some rare cases, it seems like a good idea to separate it for different test cases by clearing the buffer.

File: test/app-luatest/console_debugger_session_test.lua (hunk 5 above)
Comment (Lord-KA) [reply test/app-luatest/console_debugger_session_test.lua:330]: The comment explains what is ment by 'chunk' - one of the followup: prompt, cmd, result.

File: third_party/lua/luadebug.lua (hunk 2 above)
Comment (Lord-KA) [reply third_party/lua/luadebug.lua:30]: Yep, fixed.

File: test/app-luatest/console_debugger_session_test.lua (hunk 3 above)
Comment (Lord-KA) [reply test/app-luatest/console_debugger_session_test.lua:324]: Fixed.

File: test/app-luatest/console_debugger_session_test.lua (hunk 5 above)
Comment (Lord-KA) [reply test/app-luatest/console_debugger_session_test.lua:330]: (I don't mind concatenating the results, but it doesn't seem necessary)

File: test/app-luatest/console_debugger_session_test.lua (hunk 6 above)
Comment (Lord-KA) [reply test/app-luatest/console_debugger_session_test.lua:301]: Done.
>  The according test had to be ammended due to flaks on macOS.

File: third_party/lua/luadebug.lua (hunk 7 above)
Comment (Lord-KA) [reply third_party/lua/luadebug.lua:103]: Fixed:
>   Implemented readline history and autocomplete by reusing
  readline facilities of Tarantool console. They used to be
  being hidden once Lua 'console' module is loaded. With c432e9e 
  (lua: don't use public module name as internal one), now they 
  are available as 'console.lib'.

File: test/app-luatest/console_debugger_session_test.lua (hunk 4 above)
Comment (Totktonada) [reply test/app-luatest/console_debugger_session_test.lua:332]: Answering me here doesn't make the idea obvious from the code. Please, address such review comments in the patchset itself.

However, I'm not satisfied by the answer. As for me, it is wrong in two ways:

1. If some output is produced by a test case that is expected to don't produce any output, the problem will be hidden (instead of failing the next test case). At least, I see no check that this `:read()` call is timed out.
2. `read()` system call (and, so, our `:read()` from the pipe) has no guarantee like 'it returns everything that were written on the other side'. If two bytes were written, we can receive just one. So, in fact, it is 'usually cleans the pipe buffer, but sometimes doesn't'. It is common for byte streams. Attempt to work with a byte stream like it is a message stream is wrong.

My proposals:

* Let the next test case fail. (And properly comment how it works.)
* Trigger some response that is guaranteed to be written after the response that is expected to be empty.

File: test/app-luatest/console_debugger_session_test.lua (hunk 5 above)
Comment (Totktonada) [reply test/app-luatest/console_debugger_session_test.lua]: This is the behavior that you observe on some particular setup. It is not guaranteed by `read()`. I would prefer to lean on guarantees, especially when the work is about a test case stabilization.

If you're interested, whether it matters in practice: yes, it does. I've observed reading from a pipe with randomly splitted chunks during work on the interactive_tarantool.lua helper.

File: test/app-luatest/console_debugger_session_test.lua (hunk 6 above)
Comment (Totktonada) [reply test/app-luatest/console_debugger_session_test.lua:301]: Does it mean that the test starts to sporadically fail on macOS after the given changes in the code? I can't understand it from the comment. If it is so, I'm OK to place it into the same commit. However, if we would stabilize the test case first, it would look better, IMHO.

`checkpatch` doesn't impose on the `NO_TEST` annotation due to these changes. However, we know that the new behavior is left untested. I don't insist on testing it (it is cumbersome), but, please, leave the annotation with a reason.

Nit: there is no noun 'flak', at least it is not the same as adjective 'flaky'.

File: changelogs/unreleased/gh-7593-console-debugger.md (hunk 1 above)
Comment (Totktonada) [reply changelogs/unreleased/gh-7593-console-debugger.md]: It looks irrelevant now. I'm closing the discussion.

File: test/app-luatest/console_debugger_session_test.lua (hunk 4 above)
Comment (Lord-KA) [reply test/app-luatest/console_debugger_session_test.lua]: I rewrote the test using updated `interactive_tarantool`.

File: test/app-luatest/console_debugger_session_test.lua (hunk 6 above)
Comment (Lord-KA) [reply test/app-luatest/console_debugger_session_test.lua]: Removed the line and moved all test changes in separate commit.

File: test/app-luatest/console_debugger_session_test.lua [hunk 8]
```diff
@@ -290,36 +279,21 @@ local function run_debug_session(cmdline, sequence, tmpfile, header)
...
-            local errout = fh:read({ timeout = 0.05, stderr = true})
-            debuglog('stderr output: ', trim(errout))
-            if first and errout then
-                -- we do not expect anything on stderr
-                -- with exception of initial debugger header
-                t.assert_str_contains(trim(errout), header, false)
-                first = false
-            else
-                t.assert(errout == nil or trim(errout) == '')
-            end
```
Comment (Totktonada) [inline test/app-luatest/console_debugger_session_test.lua:313]: This commit is marked as refactoring, but actually it eliminates the stderr check. I don't know, whether it is important, but it should be mentioned in the commit message at least. If it is not important, the commit message should describe a reason.

File: test/interactive_tarantool.lua [hunk 9]
```diff
@@ -94,6 +98,23 @@ function mt.read_until_line(self, exp_line, opts)
     return res
 end
 
+-- Returns all read lines (excluding one with expected prompt).
+function mt.read_until_prompt(self, exp_prompt, opts)
```
Comment (Totktonada) [inline test/interactive_tarantool.lua:102]: Can't we use `self.prompt` here?

File: test/interactive_tarantool.lua [hunk 10]
```diff
@@ -263,18 +288,29 @@ function M.new(opts)
     local res = setmetatable({
         ph = ph,
         _readahead_buffer = '',
-        _prompt = 'tarantool> ',
+        _prompt = opts.prompt or 'tarantool> ',
```
Comment (Totktonada) [inline test/interactive_tarantool.lua:291]: Nit: I would place `local prompt = opts.prompt` next to `local args = opts.args` for consistency and to have a list of all the options at glance on the function beginning.

File: test/interactive_tarantool.lua [hunk 11]
```diff
@@ -263,18 +288,29 @@ function M.new(opts)
...
     }, mt)
 
     -- Log child's stderr.
     res:_start_stderr_logger()
 
+    return res
+end
+
+function M.new_debugger(opts)
+    opts.prompt = opts.prompt or 'luadebug> '
```
Comment (Totktonada) [inline test/interactive_tarantool.lua:301]: Nit: I would add `local opts = opts or {}` like in `_new_internal()` to allow to call this function without arguments (if no options are needed).

File: test/interactive_tarantool.lua [hunk 12]
```diff
@@ -138,11 +159,18 @@ end
...
+-- compare them for equality.
 local function _prepare_command_for_compare(x)
-    x = x:gsub(' ', ''):gsub(M.CR, '')
+    -- The terrifying regexp in the last gsub is meant to clean ANSI
+    -- color codes, for more details on the format see the doc:
+    -- https://gist.github.com/fnky/458719343aabd01cfb17a3a4f7296797
+    x = x:gsub(' ', '')
+         :gsub(M.CR, '')
+         :gsub(M.TAB, '')
+         :gsub('[\x1b\x9b][][()#;?%d]*[A-PRZcf-ntqry=><~]', '')
```
Comment (Totktonada) [inline test/interactive_tarantool.lua:173]: If it is just about colors, it should be much simpler. According to the link something like `:gsub(M.ESC .. '%[[0-9;]*m', '')` should work. If this is not only about the colors, the comment seems misleading.

Comment (Totktonada) [reply test/interactive_tarantool.lua:173]: BTW, I would put such a regexp into the exposed module constants (close to M.EXP and others). It can be useful outside of the module itself.

File: test/app-luatest/console_debugger_session_test.lua [hunk 13]
```diff
@@ -290,36 +279,21 @@ local function run_debug_session(cmdline, sequence, tmpfile, header)
...
-            repeat
-                result = trim(unescape(fh:read({ timeout = 0.5 })))
-                debuglog('stdout output:', result)
-            until result ~= '' or result ~= clean_cmd
-            if expected ~= '' then
-                t.assert_str_contains(result, expected, false)
+            if expected ~= '<END_OF_EXECUTION>' then
```
Comment (Totktonada) [inline test/app-luatest/console_debugger_session_test.lua:288]: Shouldn't we expect EOF in the `else` branch?

File: test/app-luatest/console_debugger_session_test.lua [hunk 14]
```diff
@@ -290,36 +279,21 @@ local function run_debug_session(cmdline, sequence, tmpfile, header)
...
+                local res = trim(unescape(child:read_until_prompt(dbg_prompt)))
+                debuglog(res)
+                t.assert_str_contains(res, expected)
             end
         end
-        fh:close()
+        if child ~= nil then
+            child:close()
+        end
```
Comment (Totktonada) [inline test/app-luatest/console_debugger_session_test.lua:296]: If I don't miss anything, `child` can't be `nil` here.

File: test/app-luatest/console_debugger_session_test.lua (hunk 13 above)
Comment (Lord-KA) [reply test/app-luatest/console_debugger_session_test.lua:288]: Is there a way to expect EOF with the current `it` helpers?

File: test/interactive_tarantool.lua (hunk 9 above)
Comment (Lord-KA) [reply test/interactive_tarantool.lua:102]: Ok.

File: test/interactive_tarantool.lua (hunk 10 above)
Comment (Lord-KA) [reply test/interactive_tarantool.lua:291]: It doesn't seem necessary, but I am not in the mood to argue about it.

File: test/interactive_tarantool.lua (hunk 11 above)
Comment (Lord-KA) [reply test/interactive_tarantool.lua:301]: Done.

File: test/app-luatest/console_debugger_session_test.lua (hunk 8 above)
Comment (Lord-KA) [reply test/app-luatest/console_debugger_session_test.lua:313]: Added a remark in commit msg.

File: test/app-luatest/console_debugger_session_test.lua (hunk 14 above)
Comment (Lord-KA) [reply test/app-luatest/console_debugger_session_test.lua:296]: Removed the check.

File: test/interactive_tarantool.lua (hunk 12 above)
Comment (Lord-KA) [reply test/interactive_tarantool.lua:173]: Good idea! Moved the `:gsub()` to `M.DECOLOR_RE`.

**Comment by Lord-KA** [comment]
I've reworked a bit historyfile saving, now it is being saved to `$HOME/.tdbg_history` and only when `$HOME` env var is set.

File: test/app-luatest/console_debugger_session_test.lua (hunk 8 above)
Comment (Totktonada) [reply test/app-luatest/console_debugger_session_test.lua:313]: OK. I'm relying on your investigation here.

File: test/interactive_tarantool.lua (hunk 12 above)
Comment (Totktonada) [reply test/interactive_tarantool.lua]: It looks much more accurate now.

Nit: I'm OK with regexp `M.DECOLOR_RE` or function `M.decolor`. But, please name/place it in sync with other functions/constants.

Comment (Lord-KA) [reply test/interactive_tarantool.lua]: Oh, my bad, will fix.

**Review by Totktonada — APPROVED** [review]
Visually I see no problems in the patchset.

Thank you for refactoring the test!

File: test/interactive_tarantool.lua (hunk 12 above)
Comment (Lord-KA) [reply test/interactive_tarantool.lua]: Fixed.

File: test/app-luatest/console_debugger_session_test.lua (hunk 13 above)
Comment (Lord-KA) [reply test/app-luatest/console_debugger_session_test.lua:288]: Added `t.assert_error_msg_contains()` to check expected EOF.

File: test/app-luatest/console_debugger_session_test.lua [hunk 15]
```diff
@@ -1,7 +1,6 @@
 local t = require('luatest')
```
Comment (ligurio) [inline test/app-luatest/console_debugger_session_test.lua:1]: commit "test: refactor dbgr test with it helper" breaks tests and next commit fixes it. please mention it in commit message

**Review by ligurio — COMMENTED** [review]
Gleb, thanks for the patches! see my comments

File: test/interactive_tarantool.lua [hunk 16]
```diff
@@ -138,11 +168,16 @@ end
...
--- This function drop duplicates that goes in row, strips <CR> and
--- spaces. Applying of this function to a source command and
--- readline's echoed command allows to compare them for equality.
+-- This function drop duplicates that goes in row, strips <CR>,
+-- spaces, tabs and ANSI color codes. Applying of this function
+-- to a source command and readline's echoed command allows to
+-- compare them for equality.
 local function _prepare_command_for_compare(x)
-    x = x:gsub(' ', ''):gsub(M.CR, '')
+    x = x:gsub(' ', '')
```
[the remark this answers is not in the document]
Comment (ligurio) [reply test/interactive_tarantool.lua:176]: I believe our goal is not preserving original style but producing patches with good quality.

File: test/interactive_tarantool.lua [hunk 17]
```diff
@@ -119,6 +141,14 @@ function mt.assert_data(self, exp_data, opts)
     end
 end
 
+local function decolor(string)
+    assert(type(string) == 'string')
+    -- The gsub is meant to clean ANSI color codes, for more details on
+    -- the format see the doc:
+    -- https://gist.github.com/fnky/458719343aabd01cfb17a3a4f7296797
```
[the remark this answers is not in the document]
Comment (Lord-KA) [reply test/interactive_tarantool.lua:148]: Ok, replaced the doc with [1].

File: test/app-luatest/console_debugger_session_test.lua [hunk 18]
```diff
@@ -11,17 +10,7 @@ local function trim(s)
...
 
-local function tarantool_path(arg)
-    local index = -2
-    -- arg[-1] is guaranteed to be non-null
-    while arg[index] do index = index - 1 end
-    return arg[index + 1]
-end
-
-local TARANTOOL_PATH = tarantool_path(arg)
-
```
[the remark this answers is not in the document]
Comment (Lord-KA) [reply test/app-luatest/console_debugger_session_test.lua:22]: It is a part of stderr check removal.

File: test/interactive_tarantool.lua [hunk 19]
```diff
@@ -94,6 +98,24 @@ function mt.read_until_line(self, exp_line, opts)
     return res
 end
 
+-- Returns all read lines (excluding one with expected prompt).
```
[the remark this answers is not in the document]
Comment (Lord-KA) [reply test/interactive_tarantool.lua:101]: Ok...

File: test/interactive_tarantool.lua [hunk 20]
```diff
@@ -263,18 +296,30 @@ function M.new(opts)
...
+
+function M.new(opts)
+    local res = M._new_internal(opts)
+
     -- Write a command and ignore the echoed output.
     --
     -- ReadLine 6 writes <ESC>[?1034h at beginning, it may hit
     -- assertions on the child's output.
     --
-    -- This sequence of charcters is smm ('set meta mode')
```
[the remark this answers is not in the document]
Comment (Lord-KA) [reply test/interactive_tarantool.lua:277]: >   * now there are several less tipos in comments

File: test/app-luatest/console_debugger_session_test.lua (hunk 15 above)
Comment (Lord-KA) [reply test/app-luatest/console_debugger_session_test.lua:1]: Done:
>   This patch breaks the test without debugger readline support patch,
  that is next in the patchset.

File: test/app-luatest/console_debugger_session_test.lua [hunk 21]
```diff
@@ -267,21 +256,21 @@ local sequences = {
...
     for i = 1, MAX_ALIASES_COUNT do
-        local fh = popen.new(cmdline, {
-            stdout = popen.opts.PIPE,
-            stderr = popen.opts.PIPE,
-            stdin = popen.opts.PIPE,
-        })
-        t.assert_is_not(fh, nil)
-        -- for -d option we expect debugger banner in the stderr,
-        -- but for self invoke debugger there is no extra header.
-        local first = header ~= nil
```
[the remark this answers is not in the document]
Comment (Lord-KA) [reply test/app-luatest/console_debugger_session_test.lua:284]: I removed the check because it is hard to implement using `interactive_tarantool`. `it` has stderr_logger being run in a different fiber and I couldn't get an after-test check for empty stderr working in reasonable time. As we don't expect anything from stderr, and get all the right stdout output, it seems like the test is comprehensive enough.

**Review by ylobankov — APPROVED** [review]

File: test/interactive_tarantool.lua (hunk 16 above)
[the remark this answers is not in the document]
Comment (ligurio) [reply test/interactive_tarantool.lua:176]: > Please refer to a style-guide rule. If there isn't one, it seems like a matter of taste and can be left as is.

Actually, no, it is not a matter of taste.

Style guide cannot contain all possible rules. If we try to describe all the possible rules to follow and rules that no one should break, then we will be like lawmakers, but this is not our job.

Instead of referring to a certain rule in a style guide, I will appeal to common sense and my understanding of a good programming habits, to respecting expectations of people who will read my code and use these programs.

It could be unexpected when function mutates passed arguments. Function has no any comment so without reading source code you will never know that `x` was updated. Sometimes such behavior in public functions could be a reason of real bugs, see https://github.com/tarantool/crud/issues/192.

[the remark this answers is not in the document]
Comment (Lord-KA) [reply test/interactive_tarantool.lua:176]: The bug you referred to is about a different matter. In Lua, when string is passed as an argument, like in `_prepare_command_for_compare()`, it is being copied and local variable is modified, unlike the example in https://github.com/tarantool/crud/issues/192, where a table is passed by reference.

File: test/app-luatest/console_debugger_session_test.lua (hunk 21 above)
[the remark this answers is not in the document]
Comment (Lord-KA) [reply test/app-luatest/console_debugger_session_test.lua:284]: Ok, I've had another go at this, and now it seems to be working fine. `read_until_prompt()` in `interactive_tarantool` now expects empty stderr, while `new_debugger()` expects debugger header, similar to the old test.

File: test/interactive_tarantool.lua [hunk 22]
```diff
@@ -94,6 +104,31 @@ function mt.read_until_line(self, exp_line, opts)
...
+                                                                1))
+    new_buffer = self._prompt .. new_buffer
+
+    self._readahead_buffer = new_buffer
+
+    local stderr, err = self.ph:read({timeout = 0.05, stderr = true})
+    if stderr ~= "" and not (stderr == nil and
+                             tostring(err) == "timed out") then
+        error(("Unexpected stderr output: %s"):format(stderr))
+    end
```
[the remark this answers is not in the document]
Comment (Lord-KA) [reply test/interactive_tarantool.lua:127]: No, I turned off `stderr_logger` for debugger.

[the remark this answers is not in the document]
Comment (Totktonada) [reply test/interactive_tarantool.lua:127]: I see now, thanks for the explanation.

File: test/app-luatest/console_debugger_session_test.lua (hunk 18 above)
[the remark this answers is not in the document]
Comment (ligurio) [reply test/app-luatest/console_debugger_session_test.lua:22]: right, not a dead code, it was used below

File: test/app-luatest/console_debugger_session_test.lua (hunk 21 above)
[the remark this answers is not in the document]
Comment (ligurio) [reply test/app-luatest/console_debugger_session_test.lua:284]: Thanks!

**Review by ligurio — APPROVED** [review]
LGTM, thanks, Gleb!

**Comment by igormunkin** [comment]
@Lord-KA, thanks for the changes! Could you please add the reason of reverting the changes in scope of the last patch of the series. You can see an example in 6d414e7.

**Comment by Lord-KA** [comment]
@igormunkin, done.
